zebrad/src/components/rpc.rs

```diff
@@ -0,0 +1,37 @@
+//! An RPC endpoint.
+
+use jsonrpc_core;
+
+use jsonrpc_http_server::ServerBuilder;
+
+use zebra_rpc::rpc::{Rpc, RpcImpl};
+
+pub mod config;
+pub use config::Config;
+
+/// Zebra RPC Server
+pub struct RpcServer {}
+
+impl RpcServer {
+    /// Start a new RPC server endpoint
+    pub async fn new(config: Config) -> Self {
+        if config.listen {
+            info!("Trying to open RPC endpoint at {}...", config.listen_addr);
+
+            // Create handler compatible with V1 and V2 RPC protocols
+            let mut io =
+                jsonrpc_core::IoHandler::with_compatibility(jsonrpc_core::Compatibility::Both);
+            io.extend_with(RpcImpl.to_delegate());
+
+            let server = ServerBuilder::new(io)
+                .threads(1)
+                .start_http(&config.listen_addr)
+                .expect("Unable to start RPC server");
+
+            info!("Opened RPC endpoint at {}", server.address());
+
+            server.wait();
+        }
+        RpcServer {}
```

Comment:
Blocking Problem (1/3): `server.wait()` blocks all the async tasks executing on the same thread as `RpcServer::new()`. Instead, spawn the blocking function on a blocking thread:
https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html

```suggestion
            // The server is a blocking task, so we need to spawn it on a blocking thread.
            tokio::task::spawn_blocking(server.wait)
        } else {
            // There is no RPC port, so the RPC task does nothing.
            tokio::task::spawn(futures::future::pending().in_current_span())
        }
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/3589/commits/8ed2047734f5915cc227f861d5e82991137d5ca2